wordpress-mobile/WordPress-Android
# Adding filtering feature to filterables tabs in reader

Fixes #13669 

This PR supersedes the #13728 and #13734 previous iterations. cc @malinajirka since you knows the previous iterations and we had also a pair programming session on them 🙇 (it's more like a FYI if you do not have time to have a look here, thanks 😄)

This PR introduces a filterable and organization fields to the `ReaderTag`. Currently the below are filterable:

| Tab page | Organization | Subfilter available |
|---|---|---|
| FOLLOWING | NO_ORGANIZATION | ![image](https://user-images.githubusercontent.com/47797566/104606350-f4a19e80-567f-11eb-81a4-4ea436e9043e.png) |
| FOLLOWED P2S | P2 | ![image](https://user-images.githubusercontent.com/47797566/104606682-4f3afa80-5680-11eb-8893-c6604471b8b3.png) |
| AUTOMATTIC | A8C| ![image](https://user-images.githubusercontent.com/47797566/104606627-45b19280-5680-11eb-8055-a2058d87fdcc.png) |

Note that the `TOPICS` is only available in the `FOLLOWING` bottom sheet since it makes sense only in that scope.

Moreover:
- Introduced the concept of isFilterable vs isFollowing and the functions isFilterableScreen vs isFollowingScreen (this isFollowingScreen is still used in a few places where I thought it makes sense)
- The SubfilterViewModel and the Bottom Sheet is now not shared but instantiated per ReaderPostListFragment tab page. Also the Bottom Sheet is hooked to the ReaderPostListFragment page.
- It was decided with Design to keep current filter per each tab in memory but no need to keep it when restarting the app. I reused the SubfilterListItemMap (that was used to save current filtering information into the shared prefs); in this case that is used to save into the fragment instance state. This is necessary since the fragment (and VM that is owned by it) can eventually be reinstantiated when swiping through the Reader tabs.

## To test
Please note that about unit testing, in this PR I basically fixed them to account for the changes. I will add some more coverage for the `SubFilterViewModel` in a dedicated PR.

Testing goal should be to check that everything works as before but you can additionally filter correctly not only in Following but also in P2s and A8c. Some scenarios to take care I can think of:

- User has only self hosted sites -> smoke test reader (the bottom sheet in following should offer a CTA to login)
- Fresh install (nothing in cache), create a new wp.com profile, without creating a site go to the reader and smoke test
- From the step before, create a new site and smoke test the reader
- Smoke test both from non A8C and from A8C account
- When swiping tabs far from left/rigth and vice versa check the filter selection in the 3 filterable tabs are kept. This should apply also if you place the app in background and come back to it. Restarting the app should reset the filters.
- Adding site by url (related to each of the 3 filterable tabs) you should get that new added site in the relevant subfilter when opening it

PR submission checklist:
